### ISSUE TITLE
`crane delete` requests token with insufficient scope

### ISSUE THREAD

**Issue by majewsky** [google/go-containerregistry#724]
I'm the lead developer on a cloud-hosted registry implementation called [Keppel](https://github.com/sapcc/keppel). I received a complaint from a customer that they cannot `crane delete` images stored in Keppel. I reproduced the issue; it boils down to Crane guessing the wrong token scope. When invoked as `crane delete keppel.example.com/foo/bar:qux`, it just assumes the required token scope is `repository:foo/bar:push,pull`, even though `repository:foo/bar:delete` would be required.

For reference, here's a trace of what `crane delete keppel.example.com/foo/bar:qux` does, as recorded by [`mitmdump`](https://docs.mitmproxy.org/stable/tools-mitmdump/). (The actual hostnames, repository names, and tokens have been redacted.) Note that the final request, the actual DELETE, indicates the correct token scope, but Crane ignores this information.

```
$ mitmdump -p 8888 --ssl-insecure --flow-detail 3
Proxy server listening at http://*:8888
[::1]:48300: clientconnect
[::1]:48300: GET https://keppel.example.com/v2/ HTTP/2.0
    :authority: keppel.example.com
    accept-encoding: gzip
    user-agent: Go-http-client/2.0

 << 401  104b
    server: nginx/1.15.6
    date: Wed, 27 May 2020 14:04:27 GMT
    content-type: application/json
    content-length: 104
    docker-distribution-api-version: registry/2.0
    vary: Origin
    www-authenticate: Bearer realm="https://keppel.example.com/keppel/v1/auth",service="keppel.example.com"
    strict-transport-security: max-age=15724800; includeSubDomains

    {
        "errors": [
            {
                "code": "UNAUTHORIZED",
                "detail": null,
                "message": "no bearer token found in request headers"
            }
        ]
    }

[::1]:48300: GET https://keppel.example.com/keppel/v1/auth?scope=repository%3Afoo%2Fbar%3Apush%2Cpull&service=keppel.example.com HTTP/2.0
    :authority: keppel.example.com
    authorization: Basic password-redacted
    user-agent: go-containerregistry
    accept-encoding: gzip

    scope:   repository:foo/bar:push,pull
    service: keppel.example.com

 << 200  1.65k
    server: nginx/1.15.6
    date: Wed, 27 May 2020 14:04:27 GMT
    content-type: application/json
    vary: Accept-Encoding
    vary: Origin
    strict-transport-security: max-age=15724800; includeSubDomains
    content-encoding: gzip

    {
        "expires_in": 3600,
        "issued_at": "2020-05-27T14:04:27Z",
        "token": "token-was-redacted"
    }

[::1]:48300: DELETE https://keppel.example.com/v2/foo/bar/manifests/qux HTTP/2.0
    :authority: keppel.example.com
    authorization: Bearer token-was-redacted
    user-agent: go-containerregistry
    accept-encoding: gzip

 << 403  117b
    server: nginx/1.15.6
    date: Wed, 27 May 2020 14:04:27 GMT
    content-type: application/json
    content-length: 117
    docker-distribution-api-version: registry/2.0
    vary: Origin
    www-authenticate: Bearer realm="https://keppel.example.com/keppel/v1/auth",service="keppel.example.com",scope="repository:foo/bar:delete",error="insufficient_scope"
    strict-transport-security: max-age=15724800; includeSubDomains

    {
        "errors": [
            {
                "code": "DENIED",
                "detail": null,
                "message": "token does not cover scope repository:foo/bar:delete"
            }
        ]
    }

[::1]:48300: clientdisconnect
```

**Comment by UncleIS** [comment]
I guess that here is the code the causes the problem: https://github.com/google/go-containerregistry/blob/d54baf9aa28edb9b985a6b35b57e26e3410c2443/pkg/v1/remote/transport/scope.go#L22

```
        PullScope string = "pull"
	PushScope string = "push,pull"
	// For now DELETE is PUSH, which is the read/write ACL.
	DeleteScope  string = PushScope
```

So the scope is not being derived from the URL, but is being assumed by the tool for a particular operation. I wonder if this is imposed by the API of a particular registry implementation or is that just something that was coded a while and could be adjusted?

**Comment by jonjohnsonjr** [comment]
> So the scope is not being derived from the URL, but is being assumed by the tool for a particular operation. I wonder if this is imposed by the API of a particular registry implementation or is that just something that was coded a while ago and could be adjusted?

Some rationale for why the client is generating this scope instead of relying on the registry is here:
https://github.com/containerd/containerd/issues/1318#issuecomment-327277368

tl;dr, clients usually know what they're trying to do better than the registry, so it's better for them to just generate the scope themselves.

I think it would be fine to revisit that decision, but it's tough to get this right because it's registries have wildly varying behavior. I haven't read through the documentation around scope handling in a while, so it's hard to say if this is an easy fix or not.

**Comment by jonjohnsonjr** [comment]
Use of the `delete` scope is widespread enough that it might be worth just adopting it... this was changed a while back: https://github.com/docker/distribution/commit/ccb839e0e30c3b6992fb4084dfd6550d0ddd4d1a

Annoyingly, this would break `crane delete` for GCR. We should adopt the approach of parsing the challenge and requesting the correct scope.

**Comment by jonjohnsonjr** [comment]
Slowly poking at this. I'd like to use `delete` by default here, but that will take some time to update GCR -- looking at how hard that would be.

Falling back to requesting the correct scope shouldn't be too hard either.

**Comment by jonjohnsonjr** [comment]
@majewsky is there a spec (or even just known existing behavior) somewhere around what to do if a client sends a request with insufficient scopes? Is it always 403? For your issue, I think we can just redo the token exchange if we see a 403 with a challenge that has different scopes than we sent, but I'm wondering if that's true for all registries.

**Comment by majewsky** [comment]
The distribution-spec does not say anything at all regarding AuthN/AuthZ, which this concern would fall into given that the error is communicated in the `Www-Authenticate` header. For `error="insufficient_scope"`, I was following the lead of docker-registry's implementation:

https://github.com/docker/distribution/blob/551158e6008ece74eae699e2099984d8c47393a2/registry/auth/token/accesscontroller.go#L117

> I think we can just redo the token exchange if we see a 403 with a challenge that has different scopes than we sent

That sounds like a reasonable fallback. That shouldn't break anything that's already out there.

**Comment by jonjohnsonjr** [comment]
I haven't tested this myself, but it looks like docker/distribution returns a 401 here? https://github.com/docker/distribution/blob/551158e6008ece74eae699e2099984d8c47393a2/registry/auth/token/accesscontroller.go#L95

Based on https://github.com/docker/distribution/blob/749f6afb4572201e3c37325d0ffedb6f32be8950/registry/client/auth/challenge/authchallenge.go#L132-L133 I don't expect clients to look for a challenge on a 403 status, just 401 -- is this something you could change in Keppel?

**Comment by jonjohnsonjr** [comment]
@majewsky this doesn't fix your issue because you're returning a 403, but I believe it _would_ if you returned a 401 😄 

https://github.com/google/go-containerregistry/pull/846

**Comment by majewsky** [comment]
>  is this something you could change in Keppel?

Done. See commit linked above.

**Comment by jonjohnsonjr** [comment]
Excellent! I've merged the PR that attempts to fix this -- can you confirm that `crane delete` works against Keppel with that merged?

**Comment by majewsky** [comment]
I've tested with 7c586d4a0cef22340876855603735136c846793e. It sees the 401, attempts a reauth, but with the same scope as before, so it predictably fails on second try: (once again, internal hostnames redacted)

```
$ git rev-parse HEAD
7c586d4a0cef22340876855603735136c846793e
$ go build -o ./crane ./cmd/crane
$ ./crane -v delete keppel.example.com/cc-demo/alpine:emptylayer
2020/12/01 14:52:52 --> GET https://keppel.example.com/v2/
2020/12/01 14:52:52 GET /v2/ HTTP/1.1
Host: keppel.example.com
User-Agent: crane/(devel) go-containerregistry/(devel)
Authorization: <redacted>
Accept-Encoding: gzip


2020/12/01 14:52:52 <-- 401 https://keppel.example.com/v2/ (207.154828ms)
2020/12/01 14:52:52 HTTP/2.0 401 Unauthorized
Content-Length: 104
Content-Type: application/json
Date: Tue, 01 Dec 2020 13:52:52 GMT
Docker-Distribution-Api-Version: registry/2.0
Server: nginx/1.15.6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: Origin
Www-Authenticate: Bearer realm="https://keppel.example.com/keppel/v1/auth",service="keppel.example.com"

{"errors":[{"code":"UNAUTHORIZED","message":"no bearer token found in request headers","detail":null}]}

2020/12/01 14:52:52 --> GET https://keppel.example.com/keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&service=keppel.example.com [body redacted: basic token response contains credentials]
2020/12/01 14:52:52 GET /keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&service=keppel.example.com HTTP/1.1
Host: keppel.example.com
User-Agent: crane/(devel) go-containerregistry/(devel)
Authorization: <redacted>
Accept-Encoding: gzip


2020/12/01 14:52:53 <-- 200 https://keppel.example.com/keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&service=keppel.example.com (315.457191ms) [body redacted: basic token response contains credentials]
2020/12/01 14:52:53 HTTP/2.0 200 OK
Content-Type: application/json
Date: Tue, 01 Dec 2020 13:52:53 GMT
Server: nginx/1.15.6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: Accept-Encoding
Vary: Origin


2020/12/01 14:52:53 --> DELETE https://keppel.example.com/v2/cc-demo/alpine/manifests/emptylayer
2020/12/01 14:52:53 DELETE /v2/cc-demo/alpine/manifests/emptylayer HTTP/1.1
Host: keppel.example.com
User-Agent: crane/(devel) go-containerregistry/(devel)
Authorization: <redacted>
Accept-Encoding: gzip


2020/12/01 14:52:53 <-- 401 https://keppel.example.com/v2/cc-demo/alpine/manifests/emptylayer (51.307371ms)
2020/12/01 14:52:53 HTTP/2.0 401 Unauthorized
Content-Length: 117
Content-Type: application/json
Date: Tue, 01 Dec 2020 13:52:53 GMT
Docker-Distribution-Api-Version: registry/2.0
Server: nginx/1.15.6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: Origin
Www-Authenticate: Bearer realm="https://keppel.example.com/keppel/v1/auth",service="keppel.example.com",scope="repository:cc-demo/alpine:delete",error="insufficient_scope"

{"errors":[{"code":"DENIED","message":"token does not cover scope repository:cc-demo/alpine:delete","detail":null}]}

2020/12/01 14:52:53 --> GET https://keppel.example.com/keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&scope=repository%3Acc-demo%2Falpine%3Adelete&service=keppel.example.com [body redacted: basic token response contains credentials]
2020/12/01 14:52:53 GET /keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&scope=repository%3Acc-demo%2Falpine%3Adelete&service=keppel.example.com HTTP/1.1
Host: keppel.example.com
User-Agent: crane/(devel) go-containerregistry/(devel)
Authorization: <redacted>
Accept-Encoding: gzip


2020/12/01 14:52:53 <-- 200 https://keppel.example.com/keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&scope=repository%3Acc-demo%2Falpine%3Adelete&service=keppel.example.com (243.820843ms) [body redacted: basic token response contains credentials]
2020/12/01 14:52:53 HTTP/2.0 200 OK
Content-Type: application/json
Date: Tue, 01 Dec 2020 13:52:53 GMT
Server: nginx/1.15.6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: Accept-Encoding
Vary: Origin


2020/12/01 14:52:53 --> DELETE https://keppel.example.com/v2/cc-demo/alpine/manifests/emptylayer
2020/12/01 14:52:53 DELETE /v2/cc-demo/alpine/manifests/emptylayer HTTP/1.1
Host: keppel.example.com
User-Agent: crane/(devel) go-containerregistry/(devel)
Authorization: <redacted>
Accept-Encoding: gzip


2020/12/01 14:52:53 <-- 401 https://keppel.example.com/v2/cc-demo/alpine/manifests/emptylayer (53.66226ms)
2020/12/01 14:52:53 HTTP/2.0 401 Unauthorized
Content-Length: 117
Content-Type: application/json
Date: Tue, 01 Dec 2020 13:52:53 GMT
Docker-Distribution-Api-Version: registry/2.0
Server: nginx/1.15.6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: Origin
Www-Authenticate: Bearer realm="https://keppel.example.com/keppel/v1/auth",service="keppel.example.com",scope="repository:cc-demo/alpine:delete",error="insufficient_scope"

{"errors":[{"code":"DENIED","message":"token does not cover scope repository:cc-demo/alpine:delete","detail":null}]}

2020/12/01 14:52:53 deleting keppel.example.com/cc-demo/alpine:emptylayer: DELETE https://keppel.example.com/v2/cc-demo/alpine/manifests/emptylayer: DENIED: token does not cover scope repository:cc-demo/alpine:delete
```

**Comment by jonjohnsonjr** [comment]
Here:

```
GET /keppel/v1/auth?scope=repository%3Acc-demo%2Falpine%3Apush%2Cpull&scope=repository%3Acc-demo%2Falpine%3Adelete&service=keppel.example.com HTTP/1.1
```

We are requesting both `scope=repository:cc-demo/alpine:push,pull` and `scope=repository:cc-demo/alpine:delete`.

I would expect that to work based on https://docs.docker.com/registry/spec/auth/token/

> **scope**
> The resource in question, formatted as one of the space-delimited entries from the scope parameters from the WWW-Authenticate header shown above. This query parameter should be specified multiple times if there is more than one scope entry from the WWW-Authenticate header. The above example would be specified as: scope=repository:samalba/my-app:push. The scope field may be empty to request a refresh token without providing any resource permissions to the returned bearer token.

and

> Once the token server has determined what access the client has to the resources requested in the scope parameter, it will take the intersection of the set of requested actions on each resource and the set of actions that the client has in fact been granted. If the client only has a subset of the requested access it must not be considered an error as it is not the responsibility of the token server to indicate authorization errors as part of this workflow.
> 
> Continuing with the example request, the token server will find that the client’s set of granted access to the repository is [pull, push] which when intersected with the requested access [pull, push] yields an equal set. If the granted access set was found only to be [pull] then the intersected set would only be [pull]. If the client has no access to the repository then the intersected set would be empty, [].
> 
> It is this intersected set of access which is placed in the returned token.

It might make sense for us to request `scope=repository:cc-demo/alpine:push,pull,delete` instead of both, but I think what we're sending is still valid. WDYT?

**Comment by majewsky** [comment]
> but I think what we're sending is still valid

_Well technically_ Keppel only sends exactly one scope entry in Www-Authenticate, so the sentence "This query parameter should be specified multiple times if there is more than one scope entry from the WWW-Authenticate header" would not apply. But I will still see if I can make Keppel recognize multiple scopes, at least when they concern the same resource and thus can be merged easily like in this case.

> It might make sense for us to request `scope=repository:cc-demo/alpine:push,pull,delete`

If that's an easy change on your side, I'd be in favor of that, if not for Keppel's sake, then for better compatiblity with other/future registries.

**Comment by majewsky** [comment]
I implemented the support for multiple token scopes in Keppel, and with that, `crane delete` completes successfully. From my side, this resolves the issue. @jonjohnsonjr I'll leave the decision to you if you want to close this since there was the open question whether you want to merge scopes pre-request on your side.

**Comment by jonjohnsonjr** [comment]
> I implemented the support for multiple token scopes in Keppel, and with that, crane delete completes successfully. From my side, this resolves the issue.

Thanks!

>  I'll leave the decision to you if you want to close this since there was the open question whether you want to merge scopes pre-request on your side.

I opened https://github.com/google/go-containerregistry/issues/860 to track that so that we can feel good about closing out this issue 🎉
